src/main/java/edu/kit/compiler/codegen/pattern/OperandPattern.java

```diff
@@ -66,34 +121,208 @@
-            if (node.getMode().equals(Mode.getP()) && register.isPresent()) {
-                var operand = Operand.register(node.getMode(), register.get());
-                return OperandMatch.some(Operand.memory(operand), List.of(node));
+            // We make some simplifying assumptions here. If these assumptions
+            // are not upheld, worse addressing modes may be used as a result.
+            // If arithmetic identities have been eliminated, these assumptions
+            // will always be upheld.
+            // - Constants should always be right side of addition (x + c)
+            // - Constants should always be right side of multiplication (x * c)
+            // - Subtractions of constants should not be used (x - c == x + (-c))
+
+            var nodes = AddressNodes.of(node);
+
+            var indexRight = matchMul(nodes.offset, nodes.firstRegister,
+                    nodes.secondRegister, matcher);
+            if (indexRight.matches()) {
+                return indexRight;
+            }
+
+            var indexLeft = matchMul(nodes.offset, nodes.secondRegister,
+                    nodes.firstRegister, matcher);
+            if (indexLeft.matches()) {
+                return indexLeft;
+            }
+
+            // todo (2,4,8) + 1 * %rax = (%rax,%rax,8)
+            var match = getMatch(nodes.offset, nodes.firstRegister,
+                    nodes.secondRegister, matcher);
+            if (match == null) {
+                // fallback if something goes awry
+                var register = REGISTER.match(node, matcher);
+                var operand = Operand.memory(Optional.empty(),
+                        Optional.of(register.getOperand()),
+                        Optional.empty(), Optional.empty());
+                var predecessors = register.getPredecessors()
+                        .collect(Collectors.toList());
+                return OperandMatch.some(operand, predecessors);
+            } else {
+                return match;
+            }
+        }
+
+        /**
+         * Try to match a memory addressing mode where `index` is a
+         * multiplication with one of the scaling factors 1, 2, 4, or 8.
+         */
+        private OperandMatch<Memory> matchMul(Node offset, Node base, Node index, MatcherState matcher) {
+            if (index != null && index.getOpCode() == ir_opcode.iro_Mul) {
+                var mul = (Mul) index;
+                var scaleMatch = SCALE.match(mul.getRight(), matcher);
+                if (scaleMatch.matches()) {
+                    return getMatch(Optional.ofNullable(offset),
+                            Optional.ofNullable(base), Optional.of(mul.getLeft()),
+                            Optional.of(scaleMatch.getOperand()), matcher);
+                } else {
+                    return OperandMatch.none();
+                }
+            } else {
+                return OperandMatch.none();
+            }
```

Comment:
You could eliminate one branch by using an unconditional `return` at the end of the method.

Reply:
I'm aware of that, but I try to avoid using fallthrough wherever possible, as it can make code more complex to read (obviously in this case it's pretty unnecessary).